import grpc
from concurrent import futures
import time
import io
from PIL import Image
from transformers import pipeline # Hugging Face çš„æ ¸å¿ƒåº“

import moderation_pb2
import moderation_pb2_grpc

class ModerationService(moderation_pb2_grpc.ModerationServiceServicer):
    def __init__(self):
        print("æ­£åœ¨åŠ è½½ AI æ¨¡å‹ (Falconsai/nsfw_image_detection)... ç¬¬ä¸€æ¬¡è¿è¡Œä¼šè‡ªåŠ¨ä¸‹è½½ï¼Œè¯·è€å¿ƒç­‰å¾…")
        # åˆå§‹åŒ–å›¾åƒåˆ†ç±»ç®¡é“
        # è¿™ä¸ªæ¨¡å‹ä¼šè¿”å› [{"label": "nsfw", "score": 0.99}, {"label": "normal", "score": 0.01}]
        self.classifier = pipeline("image-classification", model="Falconsai/nsfw_image_detection")
        print("AI æ¨¡å‹åŠ è½½å®Œæ¯•ï¼ŒæœåŠ¡å°±ç»ªï¼")

    def CheckImage(self, request, context):
        print(f"[Python] æ”¶åˆ°å›¾ç‰‡å®¡æ ¸è¯·æ±‚: æ–‡ä»¶å={request.file_name}, å¤§å°={len(request.image_data)} bytes")
        
        try:
            # 1. å°† gRPC ä¼ æ¥çš„ bytes è½¬æ¢ä¸º PIL å›¾ç‰‡å¯¹è±¡
            image = Image.open(io.BytesIO(request.image_data))
            
            # 2. AI æ¨ç†
            # classifier è¿”å›çš„æ˜¯ä¸€ä¸ªåˆ—è¡¨ï¼ŒåŒ…å«æ‰€æœ‰ç±»åˆ«çš„åˆ†æ•°
            results = self.classifier(image)
            # ç»“æœç¤ºä¾‹: [{'label': 'nsfw', 'score': 0.98}, {'label': 'normal', 'score': 0.02}]
            
            # 3. è§£æç»“æœ
            is_nsfw = False
            score = 0.0
            
            # æ‰¾åˆ° nsfw æ ‡ç­¾çš„åˆ†æ•°
            for res in results:
                if res['label'] == 'nsfw':
                    score = res['score']
                    break
            
            # 4. åˆ¤å®šé€»è¾‘ï¼šå¦‚æœ NSFW æ¦‚ç‡å¤§äº 80% (0.8)ï¼Œåˆ™æ‹¦æˆª
            threshold = 0.8
            if score > threshold:
                reason = f"AI æ£€æµ‹ä¸ºæ•æ„Ÿå†…å®¹ (ç½®ä¿¡åº¦: {score:.2%})"
                print(f"æ‹¦æˆªè¿è§„å›¾ç‰‡: {request.file_name}, åŸå› : {reason}")
                return moderation_pb2.CheckResponse(is_safe=False, reason=reason)
            else:
                print(f"å›¾ç‰‡é€šè¿‡: {request.file_name} (NSFW æ¦‚ç‡: {score:.2%})")
                return moderation_pb2.CheckResponse(is_safe=True, reason="Pass")

        except Exception as e:
            print(f"å¤„ç†å‡ºé”™: {e}")
            # å¦‚æœå›¾ç‰‡æŸåæˆ–æ— æ³•è¯†åˆ«ï¼Œä¸ºäº†ä¸šåŠ¡è¿ç»­æ€§ï¼Œé€šå¸¸å¯ä»¥é€‰æ‹©æ”¾è¡Œæˆ–æŠ¥é”™
            # è¿™é‡Œæ¼”ç¤ºæŠ¥é”™
            return moderation_pb2.CheckResponse(is_safe=False, reason=f"å›¾ç‰‡å¤„ç†å¤±è´¥: {str(e)}")

def serve():
    # å®šä¹‰æœ€å¤§æ¶ˆæ¯å¤§å° (ä¾‹å¦‚ 50MB = 50 * 1024 * 1024)
    MAX_MESSAGE_LENGTH = 50 * 1024 * 1024

    # å»ºç«‹ gRPC æœåŠ¡å™¨æ—¶ï¼Œä¼ å…¥ options å‚æ•°
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=10),
        options=[
            ('grpc.max_send_message_length', MAX_MESSAGE_LENGTH),
            ('grpc.max_receive_message_length', MAX_MESSAGE_LENGTH),
        ]
    )
    
    # å®ä¾‹åŒ–æœåŠ¡ç±»
    service_instance = ModerationService()
    
    moderation_pb2_grpc.add_ModerationServiceServicer_to_server(service_instance, server)
    
    server.add_insecure_port('[::]:50051')
    print(f"ğŸš€ Python AI å†…å®¹å®¡æ ¸å¾®æœåŠ¡å·²å¯åŠ¨ (Port: 50051) | æœ€å¤§æ¶ˆæ¯é™åˆ¶: {MAX_MESSAGE_LENGTH/1024/1024}MB...")
    try:
        server.start()
        while True:
            time.sleep(86400)
    except KeyboardInterrupt:
        server.stop(0)

if __name__ == '__main__':
    serve()